pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -234,204 +234,283 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, ProtocolFeeCache {
-        uint256 bptIn,
-        uint256 tokenIndex,
+    function _onSwapBptGivenIn(
+        uint256 amount,
+        uint256 indexIn,
+        uint256 indexOut,
+        bool bptIsTokenIn,
+        uint256 amp,
+        uint256 protocolSwapFeePercentage,
         uint256 virtualSupply,
-        uint256[] memory balances
-    ) internal view returns (uint256 amountOut) {
-        // Use virtual total supply and zero swap fees for joins.
-        (uint256 amp, ) = _getAmplificationParameter();
-        amountOut = StableMath._calcTokenOutGivenExactBptIn(amp, balances, tokenIndex, bptIn, virtualSupply, 0);
-    }
+        uint256[] memory balancesWithoutBpt,
+        uint256[] memory scalingFactors
+    ) private returns (uint256) {
+        uint256 swapFeePercentage = getSwapFeePercentage();
+        uint256 amountOut;
+
+        if (bptIsTokenIn) {
+            // exitSwap
+            amountOut = StableMath._calcTokenOutGivenExactBptIn(
+                amp,
+                balancesWithoutBpt,
+                _skipBptIndex(indexOut),
+                amount,
+                virtualSupply,
+                swapFeePercentage
+            );
+        } else {
+            // joinSwap
+            uint256[] memory amountsIn = new uint256[](_getTotalTokens() - 1);
+            amountsIn[_skipBptIndex(indexIn)] = _upscale(amount, scalingFactors[indexIn]);
 
-    /**
-     * @dev Calculate token in for exact BPT out (join)
-     */
-    function _onSwapTokenGivenBptOut(
-        uint256 bptOut,
-        uint256 tokenIndex,
-        uint256 virtualSupply,
-        uint256[] memory balances
-    ) internal view returns (uint256 amountIn) {
-        // Use virtual total supply and zero swap fees for joins
-        (uint256 amp, ) = _getAmplificationParameter();
-        amountIn = StableMath._calcTokenInGivenExactBptOut(amp, balances, tokenIndex, bptOut, virtualSupply, 0);
-    }
+            amountOut = StableMath._calcBptOutGivenExactTokensIn(
+                amp,
+                balancesWithoutBpt,
+                amountsIn,
+                virtualSupply,
+                swapFeePercentage
+            );
+        }
 
-    /**
-     * @dev Calculate BPT in for exact token out (exit)
-     */
-    function _onSwapBptGivenTokenOut(
-        uint256 amountOut,
-        uint256 tokenIndex,
-        uint256 virtualSupply,
-        uint256[] memory balances
-    ) internal view returns (uint256 bptIn) {
-        // Avoid BPT balance for stable pool math. Use virtual total supply and zero swap fees for exits.
-        (uint256 amp, ) = _getAmplificationParameter();
-        uint256[] memory amountsOut = new uint256[](_getTotalTokens() - 1);
-        amountsOut[tokenIndex] = amountOut;
-        bptIn = StableMath._calcBptInGivenExactTokensOut(amp, balances, amountsOut, virtualSupply, 0);
+        if (protocolSwapFeePercentage > 0) {
+            _payDueProtocolFeeByBpt(bptIsTokenIn ? amount : amountOut, protocolSwapFeePercentage);
```

Comment:
also unscaled here - see the suggestion on how to fix this above

Reply:
It only scales when it needs to, taking advantage of knowing when it's BPT or not